Procedures to suspend and activate users

fickludd opened the pull request.

fickludd committed:
Started working on suspending users.

Boolean impl on User.

craigtaverner committed:
Enabled suspend/activate user and fixed tests

fickludd committed:
Completion of suspend / activate user.

* general purpose flags on User class and it's serialization
* cleaned up User builder
* added UserTest on immutability
* improved AuthProcedureTests on suspend / activate user

OliviaYtterbrink commented on enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/AuthProceduresTest.java:
```diff
@@ -391,21 +391,86 @@ public void roleManagement2() throws Exception
+    Admin suspends user Henrik
+    Henrik’s session is terminated
+    Henrik logs in with correct password → fail
+     */
```
This comment should reflect the scenario (2) in the use case document.

henriknyman commented on community/security/src/main/java/org/neo4j/server/security/auth/User.java:
```diff
@@ -35,14 +39,16 @@
+    public static final String PASSWORD_CHANGE_REQUIRED = "password_change_required";
 
-    public User(String name, Credential credential, boolean passwordChangeRequired)
+    private User( String name, Credential credential, HashSet<String> flags )
```
Maybe use an EnumSet instead?

lutovich commented on community/security/src/main/java/org/neo4j/server/security/auth/User.java:
```diff
@@ -35,14 +39,16 @@
-    /** Whether a password change is needed */
-    private final boolean passwordChangeRequired;
+    /** Set of flags, eg. require_password_change */
+    private final HashSet<String> flags;
```
It is better to use `Set` interface here and in constructor.

henriknyman commented on community/security/src/main/java/org/neo4j/server/security/auth/User.java:
```diff
@@ -105,32 +114,69 @@ public String toString()
         }
     }
+
+    private static boolean setsAreEquals( Set<String> set1, Set<String> set2 )
```
Can't we just use the default equals() instead of this?

lutovich commented on community/security/src/main/java/org/neo4j/server/security/auth/User.java:
```diff
@@ -105,32 +114,69 @@ public String toString()
         private String name;
         private Credential credential = Credential.INACCESSIBLE;
-        private boolean pwdChangeRequired;
+        private HashSet<String> flags = new HashSet<>();
```
Please use `Set` in the left hand side

lutovich commented on community/security/src/main/java/org/neo4j/server/security/auth/UserSerialization.java:
```diff
@@ -115,11 +129,18 @@ private Credential deserializeCredentials( String part, int lineNumber ) throws
+        return join(separator, Arrays.asList(segments).iterator() );
+    }
+
+    private String join( String separator, Iterator<String> segments )
```
Please consider turning `segments` into stream and using joining collector instead of writing join logic once again.

OliviaYtterbrink replied:
An alternative implementation: 

```
StringBuilder sb = new StringBuilder();
String delim = "";
while (segments.hasNext())
{
    sb.append( delim + segments.next() );
    delim = separator;
}
return sb.toString();
```

lutovich replied:
Let's just use `String#join(CharSequence delimiter, Iterable<? extends CharSequence> elements)`

lutovich commented on community/security/src/main/java/org/neo4j/server/security/auth/UserSerialization.java:
```diff
@@ -115,11 +129,18 @@ private Credential deserializeCredentials( String part, int lineNumber ) throws
 
     private String join( String separator, String... segments )
     {
+        return join(separator, Arrays.asList(segments).iterator() );
```
Maybe use `Iterators#iterator(segments)`

henriknyman commented on enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroAuthManager.java:
```diff
@@ -160,14 +161,10 @@ public AuthSubject login( String username, String password )
+                if ( realm.findUser( username ).passwordChangeRequired() )
+                {
+                    result = AuthenticationResult.PASSWORD_CHANGE_REQUIRED;
+                }
```
This moves us slightly away from multi-realm authentication, since it is assuming the login was handled by FileUserRealm

OliviaYtterbrink commented on enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/AuthProceduresTest.java:
```diff
@@ -391,21 +391,86 @@ public void roleManagement2() throws Exception
         testSuccessfulReadAction( subject, 4L );
     }
 
+    //----------User suspension scenarios-----------
```
Unit tests are missing for activating and suspending users. At least the following tests need to be added:
- Activating an already activated user
- Suspending an already suspended user
- Non-admin activating/suspending other user

lutovich commented:
@fickludd code in this PR does not adhere to our code style. please format all touched files.

henriknyman commented on enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/FileUserRealm.java:
```diff
@@ -106,9 +108,15 @@ public FileUserRealm( UserRepository userRepository, RoleRepository roleReposito
-
-        Set<String> roles = roleRepository.findByUsername( user.name() );
-        return new SimpleAuthorizationInfo( roles );
+        if ( user.passwordChangeRequired() || user.hasFlag( IS_SUSPENDED ))
```
How about a user.isSuspended() helper?

OliviaYtterbrink replied:
Have `user.isSuspended()` instead of checking the flag here so as not to leak implementation details.

henriknyman commented on community/security/src/main/java/org/neo4j/server/security/auth/UserSerialization.java:
```diff
@@ -115,11 +129,18 @@ private Credential deserializeCredentials( String part, int lineNumber ) throws
+    }
+
+    private String join( String separator, Iterator<String> segments )
+    {
```
I think maybe we should guarantee the sort order of flags when serializing (like we do with roles)

fickludd committed:
FIxed test on UserServiceTest to use new User creation

OliviaYtterbrink commented on enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/FileUserRealm.java:
```diff
@@ -57,6 +57,8 @@
     private final UserRepository userRepository;
     private final RoleRepository roleRepository;
 
+    public static final String IS_SUSPENDED = "is_suspended";
```
This should exist in User.java instead along with PASSWORD_CHANGE_REQUIRED

OliviaYtterbrink commented on enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/AuthProcedures.java:
```diff
@@ -74,4 +75,28 @@ public void removeUserFromRole( @Name( "username" ) String username, @Name( "rol
+
+    @PerformsDBMS
+    @Procedure( "dbms.suspendUser" )
+    public void suspendUser( @Name( "username" ) String username ) throws IOException, ConcurrentModificationException
```
We don't want to expose the `ConcurrentModificationException` in procedures (see comments in `FileUserRealm`).

OliviaYtterbrink commented on enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/AuthProcedures.java:
```diff
@@ -74,4 +75,28 @@ public void removeUserFromRole( @Name( "username" ) String username, @Name( "rol
+
+    @PerformsDBMS
+    @Procedure( "dbms.activateUser" )
+    public void activateUser( @Name( "username" ) String username ) throws IOException, ConcurrentModificationException
```
We don't want to expose the ConcurrentModificationException in procedures (see comments in FileUserRealm).

OliviaYtterbrink commented on enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/FileUserRealm.java:
```diff
@@ -260,6 +268,31 @@ boolean deleteUser( String username ) throws IOException
         return result;
     }
 
+    void suspendUser( String username ) throws IOException, ConcurrentModificationException
```
This method isn't properly synchronized and `ConcurrentModificationException` shouldn't be bubbled up. (See methods above)

OliviaYtterbrink commented on enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/FileUserRealm.java:
```diff
@@ -260,6 +268,31 @@ boolean deleteUser( String username ) throws IOException
+        }
+    }
+
+    void activateUser( String username ) throws IOException, ConcurrentModificationException
```
This method isn't properly synchronized and `ConcurrentModificationException` shouldn't be bubbled up. (See methods above)

OliviaYtterbrink commented on enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroAuthManager.java:
```diff
@@ -216,6 +213,18 @@ public boolean deleteUser( String username ) throws IOException
         return realm.deleteUser( username );
     }
 
+    void suspendUser( String username ) throws IOException, ConcurrentModificationException
```
We don't want to expose the `ConcurrentModificationException` (see comments in FileUserRealm).

OliviaYtterbrink commented on enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroAuthManager.java:
```diff
@@ -216,6 +213,18 @@ public boolean deleteUser( String username ) throws IOException
+        realm.suspendUser( username );
+    }
+
+    void activateUser( String username ) throws IOException, ConcurrentModificationException
```
We don't want to expose the `ConcurrentModificationException` (see comments in FileUserRealm).

fickludd committed:
Improved suspend/activate based on PR feedback.

henriknyman commented on enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/FileUserRealm.java:
```diff
@@ -270,6 +283,63 @@ boolean deleteUser( String username ) throws IOException
+                {
+                    // Due to the synchronized (this) we should never get this exception unless the user repository
+                    // is modified outside this neo instance. In that case we do not know how to proceed.
+                    throw new RuntimeException( "Unexpected failure", e );
```
I think we should follow the same pattern as with all the other methods and just retry (*unconditionally) and not introduce unnecessary assumptions about how this repository is being used at this point. (*The likelihood of an endless race between multiple entities modifying the same repository record concurrently should be reasonably slim)

OliviaYtterbrink commented on enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/AuthProceduresTest.java:
```diff
@@ -544,6 +547,72 @@ public void userDeletion4() throws Exception
         }
     }
 
+    //----------User suspension scenarios-----------
```
Unit tests are missing for activating and suspending users from the _procedures_ point of view. At least the following tests need to be added:

Activating an already activated user
Suspending an already suspended user
Non-admin activating/suspending other user

fickludd committed:
Added unit tests for suspend/activate user at procedure level

Changed suspendUser and activateUser synchronization in FileUserRealm
to match previous solution.
